src/inmanta/util.py

```diff
@@ -236,14 +236,18 @@ def __init__(self, name: str) -> None:
         # Keep track of all tasks that are currently executing to be
         # able to cancel them when the scheduler is stopped.
         self._executing_tasks: Dict[TaskMethod, List[asyncio.Task[object]]] = defaultdict(list)
+        # Keep track of tasks that should be awaited before the scheduler is stopped
+        self._await_tasks: Dict[TaskMethod, List[asyncio.Task[object]]] = defaultdict(list)
 
-    def _add_to_executing_tasks(self, action: TaskMethod, task: asyncio.Task[object]) -> None:
+    def _add_to_executing_tasks(self, action: TaskMethod, task: asyncio.Task[object], cancel_on_stop: bool = True) -> None:
         """
         Add task that is currently executing to `self._executing_tasks`.
         """
         if action in self._executing_tasks and self._executing_tasks[action]:
             LOGGER.warning("Multiple instances of background task %s are executing simultaneously", action.__name__)
         self._executing_tasks[action].append(task)
+        if not cancel_on_stop:
+            self._await_tasks[action].append(task)
```

Comment:
The tasks in the `_await_tasks` dictionary are not properly cleaned up when the task finishes execution. As such, they will grow until infinity.